clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java

```diff
@@ -1,33 +1,166 @@
+    protected static final Logger LOG = LoggerFactory.getLogger(LakeFSFileSystemTest.class);
+    protected final LakeFSFileSystem fs = new LakeFSFileSystem();
+
+    protected LakeFSClient lfsClient;
+    protected ObjectsApi objectsApi;
+    protected RepositoriesApi repositoriesApi;
+    protected StagingApi stagingApi;
+
+    protected AmazonS3 s3Client;
 
-    private LakeFSFileSystem fs;
+    protected String s3Base;
+    protected String s3Bucket;
+
+    private static final DockerImageName MINIO = DockerImageName.parse("minio/minio:RELEASE.2021-05-16T05-32-34Z");
+    protected static final String S3_ACCESS_KEY_ID = "AKIArootkey";
+    protected static final String S3_SECRET_ACCESS_KEY = "secret/minio/key=";
+
+    protected static final Regions REGION = Regions.US_EAST_1;
+
+    @Rule
+    public final GenericContainer s3 = new GenericContainer(MINIO.toString()).
+        withCommand("minio", "server", "/data").
+        withEnv("MINIO_ROOT_USER", S3_ACCESS_KEY_ID).
+        withEnv("MINIO_ROOT_PASSWORD", S3_SECRET_ACCESS_KEY).
+        //        withEnv("MINIO_REGION_NAME", REGION.getName()).
+        withEnv("MINIO_DOMAIN", "s3.local.lakefs.io").
+        withExposedPorts(9000);
+
+    protected static String makeS3BucketName() {
+        String slug = NanoIdUtils.randomNanoId(NanoIdUtils.DEFAULT_NUMBER_GENERATOR,
+                                               "abcdefghijklmnopqrstuvwxyz-0123456789".toCharArray(), 14);
+        return String.format("bucket-%s-x", slug);
+    }
+
+    /** @return "s3://..." URL to use for s3Path (which does not start with a slash) on bucket */
+    protected String s3Url(String s3Path) {
+        return s3Base + s3Path;
+    }
+
+    @Before
+    public void logS3Container() {
+        Logger s3Logger = LoggerFactory.getLogger("s3 container");
+        Slf4jLogConsumer logConsumer = new Slf4jLogConsumer(s3Logger).
+            withMdc("container", "s3").
+            withSeparateOutputStreams();
+        s3.followOutput(logConsumer);
+    }
 
     @Before
     public void setUp() throws Exception {
-        fs = new LakeFSFileSystem();
+        AWSCredentials creds = new BasicAWSCredentials(S3_ACCESS_KEY_ID, S3_SECRET_ACCESS_KEY);
+
+        ClientConfiguration clientConfiguration = new ClientConfiguration()
+                .withSignerOverride("AWSS3V4SignerType");
+        String s3Endpoint = String.format("http://s3.local.lakefs.io:%d", s3.getMappedPort(9000));
+
+        s3Client = new AmazonS3Client(creds, clientConfiguration);
+
+        S3ClientOptions s3ClientOptions = new S3ClientOptions()
+            .withPathStyleAccess(true);
+        s3Client.setS3ClientOptions(s3ClientOptions);
+        s3Client.setEndpoint(s3Endpoint);
+
+        // S3ClientOptions opts = new S3ClientOptions();
+        // opts.setPathStyleAccess(true);
+        // s3Client.setS3ClientOptions(opts);
+
+        s3Bucket = makeS3BucketName();
+        s3Base = String.format("s3://%s", s3Bucket);
```

Comment:
bucket doesn't end with '/' - so we use path with '/' prefix or we need to add one here.

Reply:
Yes, this is done throughout the test.  Not changing.